<a href="https://colab.research.google.com/github/Suchrahmadhani/DataScience-Practicum/blob/main/Pertemuan6_Suci_Rahmadhani_240401070508.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook 6: Data Preprocessing untuk Machine Learning

## Identitas Mahasiswa

| Keterangan | Informasi |
|------------|-----------|
| **Nama** | Suci Rahmadhani |
| **NIM** | 240401070508 |
| **Kelas** | IF404 |
| **Program Studi** | PJJ Informatika |
| **Mata Kuliah** | Data Science |

# Tujuan Praktikum

Pada praktikum ini saya mempelajari tahapan **data preprocessing** sebagai proses persiapan data sebelum membangun model machine learning. Praktikum meliputi eksplorasi awal dataset (Exploratory Data Analysis/EDA), penanganan *missing values*, konversi data kategorikal menjadi numerik (*encoding*), pembagian dataset menjadi data latih dan data uji (*train-test split*), serta normalisasi fitur menggunakan **StandardScaler**. Seluruh tahapan ini bertujuan agar dataset siap digunakan untuk proses pemodelan machine learning.

# Ringkasan Materi

Data preprocessing merupakan salah satu tahapan yang sangat penting dalam Data Science dan Machine Learning. Dataset mentah sering kali masih memiliki berbagai permasalahan, seperti data yang hilang (*missing values*), data kategorikal yang belum dapat diproses oleh algoritma machine learning, maupun perbedaan skala antar fitur numerik. Oleh karena itu, diperlukan proses preprocessing agar data menjadi lebih bersih, konsisten, dan siap digunakan.

Pada praktikum ini digunakan **dataset Titanic** yang berisi informasi mengenai penumpang kapal Titanic, seperti kelas penumpang, usia, jenis kelamin, tarif, pelabuhan keberangkatan, dan status keselamatan penumpang. Dataset tersebut diproses menggunakan beberapa teknik preprocessing sebelum nantinya digunakan dalam proses pembangunan model machine learning.

In [1]:
# Import Library

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Eksplorasi Awal Dataset (EDA)

Tahap pertama dalam data preprocessing adalah melakukan eksplorasi awal terhadap dataset. Pada tahap ini dilakukan pemeriksaan ukuran dataset, tipe data setiap kolom, beberapa baris pertama dataset, serta jumlah *missing values*. Informasi tersebut membantu memahami kondisi awal dataset sebelum dilakukan proses pembersihan dan persiapan data.

In [2]:
# Membaca dataset Titanic

df = pd.read_csv("titanic.csv")

print("Shape Dataset :", df.shape)

print("\nTipe Data")
display(df.dtypes)

print("\nLima Data Pertama")
display(df.head())

print("\nMissing Values")
display(df.isnull().sum())

print("\nDistribusi Target (Survived)")
display(df["Survived"].value_counts(normalize=True).round(3))

Shape Dataset : (891, 12)

Tipe Data


,0
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64



Lima Data Pertama


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S



Missing Values


,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0



Distribusi Target (Survived)


,proportion
Survived,
0,0.616
1,0.384


### Interpretasi

Hasil eksplorasi awal menunjukkan bahwa dataset Titanic terdiri dari 891 baris dan 12 kolom. Dataset ini memuat berbagai informasi mengenai karakteristik penumpang, seperti kelas penumpang, usia, jenis kelamin, tarif, pelabuhan keberangkatan, serta status keselamatan. Selain itu, masih terdapat beberapa kolom yang memiliki *missing values*, sehingga diperlukan proses preprocessing sebelum dataset digunakan pada tahap machine learning.

# Menangani Missing Values

Sebelum dataset digunakan untuk machine learning, perlu dilakukan pemeriksaan dan penanganan terhadap *missing values*. Pada praktikum ini, nilai yang hilang pada kolom numerik akan diisi menggunakan median, sedangkan kolom kategorikal akan diisi menggunakan modus (nilai yang paling sering muncul).

In [3]:
# Menampilkan jumlah missing values sebelum preprocessing

print("Missing Values Sebelum Penanganan")
display(df.isnull().sum())

Missing Values Sebelum Penanganan


,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


### Interpretasi

Hasil pemeriksaan menunjukkan bahwa dataset masih memiliki beberapa *missing values*, terutama pada kolom Age, Cabin, dan Embarked. Pada praktikum ini, hanya kolom yang digunakan untuk pemodelan yang akan ditangani sehingga proses preprocessing menjadi lebih efisien.

In [4]:
# Mengisi missing value

# Kolom numerik menggunakan median
df["Age"] = df["Age"].fillna(df["Age"].median())

# Kolom kategorikal menggunakan modus
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])

print("Missing Values Setelah Penanganan")
display(df.isnull().sum())

Missing Values Setelah Penanganan


,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,0
SibSp,0
Parch,0
Ticket,0
Fare,0


### Interpretasi

Missing value pada kolom Age berhasil diisi menggunakan nilai median sehingga tidak terlalu dipengaruhi oleh nilai ekstrem. Sementara itu, kolom Embarked diisi menggunakan nilai modus karena merupakan kategori yang paling sering muncul. Setelah proses ini selesai, kolom yang akan digunakan dalam pemodelan telah terbebas dari *missing values*.

### Catatan

Kolom **Cabin** masih memiliki banyak missing values, namun kolom tersebut tidak digunakan sebagai fitur dalam praktikum ini. Oleh karena itu, proses imputasi hanya dilakukan pada kolom Age dan Embarked yang digunakan pada tahap preprocessing dan pemodelan.

# Encoding Data Kategorikal

Sebagian algoritma machine learning hanya dapat memproses data numerik. Oleh karena itu, variabel kategorikal seperti Sex dan Embarked perlu diubah menjadi bentuk numerik menggunakan teknik One-Hot Encoding.

In [5]:
# One-Hot Encoding

df = pd.get_dummies(
    df,
    columns=["Sex", "Embarked"],
    drop_first=True,
    dtype=int
)

print("Kolom Setelah Encoding")

display(df.columns)

Kolom Setelah Encoding


Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Sex_male', 'Embarked_Q', 'Embarked_S'],
      dtype='object')

### Interpretasi

Variabel Sex dan Embarked berhasil dikonversi menjadi beberapa kolom numerik menggunakan teknik One-Hot Encoding. Penggunaan parameter drop_first=True bertujuan untuk menghindari terjadinya *dummy variable trap* sehingga model machine learning menjadi lebih stabil.

# Menentukan Fitur dan Target

Sebelum membagi dataset menjadi data latih dan data uji, terlebih dahulu ditentukan variabel target yang akan diprediksi serta fitur-fitur yang digunakan sebagai masukan model. Pada praktikum ini, variabel target yang digunakan adalah Survived.

In [6]:
# Memilih fitur yang akan digunakan

fitur = [
    "Pclass",
    "Age",
    "SibSp",
    "Parch",
    "Fare",
    "Sex_male",
    "Embarked_Q",
    "Embarked_S"
]

X = df[fitur]

y = df["Survived"]

print("Shape X :", X.shape)
print("Shape y :", y.shape)

Shape X : (891, 8)
Shape y : (891,)


### Interpretasi

Variabel Survived dipilih sebagai target karena akan digunakan untuk memprediksi apakah seorang penumpang selamat atau tidak. Sementara itu, fitur-fitur yang dipilih terdiri dari informasi mengenai kelas penumpang, usia, jumlah keluarga, tarif, jenis kelamin, dan pelabuhan keberangkatan.

# Train-Test Split

Tahap berikutnya adalah membagi dataset menjadi data latih (training set)dan data uji (testing set). Data latih digunakan untuk melatih model machine learning, sedangkan data uji digunakan untuk mengevaluasi performa model. Pada praktikum ini digunakan pembagian 80% data latih dan 20% data uji.

In [7]:
# Membagi dataset menjadi data training dan data testing

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Shape Data Training :", X_train.shape)
print("Shape Data Testing  :", X_test.shape)

print()

print("Distribusi Target Training")
display(y_train.value_counts(normalize=True).round(3))

print()

print("Distribusi Target Testing")
display(y_test.value_counts(normalize=True).round(3))

Shape Data Training : (712, 8)
Shape Data Testing  : (179, 8)

Distribusi Target Training


,proportion
Survived,
0,0.617
1,0.383



Distribusi Target Testing


,proportion
Survived,
0,0.615
1,0.385


### Interpretasi

Dataset berhasil dibagi menjadi 80% data training dan 20% data testing. Parameter **stratify=y** digunakan agar proporsi penumpang yang selamat dan tidak selamat tetap seimbang pada kedua kelompok data. Pembagian ini penting agar model dapat dilatih dan dievaluasi menggunakan data yang representatif.

# Feature Scaling

Setelah dataset dibagi menjadi data training dan data testing, dilakukan proses feature scaling menggunakan StandardScaler. Proses ini bertujuan untuk menyamakan skala setiap fitur numerik sehingga algoritma machine learning dapat bekerja dengan lebih optimal.

In [8]:
# Memilih kolom numerik yang akan dinormalisasi

numerical_features = [
    "Age",
    "Fare",
    "SibSp",
    "Parch"
]

scaler = StandardScaler()

# Fit hanya pada data training
X_train[numerical_features] = scaler.fit_transform(
    X_train[numerical_features]
)

# Transform data testing
X_test[numerical_features] = scaler.transform(
    X_test[numerical_features]
)

print("Lima Data Training Setelah StandardScaler")

display(X_train.head())

Lima Data Training Setelah StandardScaler


,Pclass,Age,SibSp,Parch,Fare,Sex_male,Embarked_Q,Embarked_S
692,3,-0.112078,-0.465084,-0.466183,0.513812,1,0,1
481,2,-0.112078,-0.465084,-0.466183,-0.662563,1,0,1
527,1,-0.112078,-0.465084,-0.466183,3.955399,1,0,1
855,3,-0.879807,-0.465084,0.727782,-0.467874,0,0,1
801,2,0.118241,0.478335,0.727782,-0.115977,0,0,1


### Interpretasi

StandardScaler berhasil mengubah skala fitur numerik sehingga memiliki rata-rata mendekati 0 dan standar deviasi mendekati 1. Proses fit hanya dilakukan pada data training, sedangkan data testing hanya menggunakan transform untuk menghindari terjadinya *data leakage*. Tahapan ini sangat penting agar model machine learning tidak dipengaruhi oleh perbedaan skala antar fitur.

# Validasi Hasil Preprocessing

Sebelum digunakan pada tahap pemodelan machine learning, dilakukan pemeriksaan akhir untuk memastikan bahwa dataset telah berhasil diproses dengan baik.

In [9]:
print("Shape X Train :", X_train.shape)
print("Shape X Test  :", X_test.shape)

print()

print("Shape y Train :", y_train.shape)
print("Shape y Test  :", y_test.shape)

print()

print("Apakah masih ada missing values?")

print(X_train.isnull().sum().sum())

print(X_test.isnull().sum().sum())

Shape X Train : (712, 8)
Shape X Test  : (179, 8)

Shape y Train : (712,)
Shape y Test  : (179,)

Apakah masih ada missing values?
0
0


### Interpretasi

Hasil validasi menunjukkan bahwa data training maupun data testing telah berhasil dipersiapkan untuk proses machine learning. Tidak terdapat lagi *missing values* pada fitur yang digunakan, seluruh variabel kategorikal telah dikonversi menjadi numerik, serta fitur numerik telah dinormalisasi menggunakan StandardScaler.

# Kesimpulan

Pada praktikum ini saya mempelajari proses **data preprocessing** menggunakan dataset Titanic sebagai persiapan sebelum membangun model machine learning. Tahapan yang dilakukan meliputi eksplorasi awal dataset, penanganan *missing values*, konversi data kategorikal menggunakan **One-Hot Encoding**, pembagian dataset menjadi data training dan data testing, serta normalisasi fitur menggunakan **StandardScaler**.

Melalui praktikum ini saya memahami bahwa data preprocessing merupakan tahapan yang sangat penting karena dapat meningkatkan kualitas data sebelum digunakan dalam proses pemodelan. Saya juga mengetahui bahwa setiap tahapan preprocessing memiliki tujuan yang berbeda, mulai dari membersihkan data hingga menyiapkan data agar dapat diproses secara optimal oleh algoritma machine learning. Pada pertemuan berikutnya, saya berharap dapat menerapkan dataset yang telah diproses ini untuk membangun dan mengevaluasi model machine learning.